packages/library/src/commonMain/kotlin/io/realm/Realm.kt

```diff
@@ -57,10 +64,54 @@ class Realm private constructor(configuration: RealmConfiguration, dbPointer: Na
         this(configuration, RealmInterop.realm_open(configuration.nativeConfig))
 
     /**
-     * TODO Add docs when this method is implemented.
+     * Modify the underlying Realm file in a suspendable transaction on the default Realm
+     * dispatcher.
+     *
+     * NOTE: Objects and results retrieved before a write are no longer valid. This restriction
+     * will be lifted when the frozen architecture is fully in place.
+     *
+     * The write transaction always represent the latest version of data in the Realm file, even if
+     * the calling Realm not yet represent this.
+     *
+     * Write transactions automatically commit any changes made when the closure returns unless
+     * [MutableRealm.cancelWrite] was called.
+     *
+     * @param block function that should be run within the context of a write transaction.
+     * FIXME Isn't this impossible to achieve in a way where we can guarantee that we freeze all
+     *  objects leaving the transaction? It currently works for RealmObjects, and can maybe be done
+     *  for collections and RealmResults, but what it is bundled in other containers, etc. Should
+     *  we define an interface of _returnable_ objects that follows some convention?
+     * @return any value returned from the provided write block as frozen/immutable objects.
+     * @see [RealmConfiguration.writeDispatcher]
      */
-    suspend fun <R> write(block: MutableRealm.() -> R): R {
-        TODO("Awaiting implementation of the Frozen Architecture")
+    // TODO Would we be able to offer a per write error handler by adding a CoroutineExceptinoHandler
+    internal suspend fun <R> write(block: MutableRealm.() -> R): R {
```

Comment:
I intentionally didn't make it suspendable because I think there is some fundamental clarity in _not_ having it. It makes most sense to avoid doing other asynchronous stuff here. We would also need to ensure that we avoid deadlocking, but I guess there is much more to it, as the live objects cannot be pass on from here since they are still thread confined.

Reply:
Yes, after thinking about this, it seems reasonable to not support suspend functions inside the write. We can always re-evaluate based on user feedback.

Reply:
I left it for now, but there are some annoyances around Darwin testing for our current setup, as tests and writer is using the same thread/dispatcher, which makes is pretty hard to script proper flows without being able to suspend from the write-block.